### PR TITLE
Update ML.NET to 1.4

### PR DESCRIPTION
**Pull request by adamsitnik**
Before I start working on #896 we need to update the version of ML.NET first

### CODE REVIEW

File: src/benchmarks/real-world/Microsoft.ML.Benchmarks/HashBench.cs [hunk 1]
```diff
@@ -20,73 +21,93 @@ namespace Microsoft.ML.Benchmarks
...
+                .GetFields(BindingFlags.NonPublic | BindingFlags.Instance)
+                .Single(field => field.Name == "_columns")
+                .GetValue(info);
+            var xf = typeof(HashingTransformer)
+                .GetConstructor(
+                    BindingFlags.NonPublic | BindingFlags.Instance, 
+                    null, 
+                    new[] { typeof(IHostEnvironment), columns.GetType() }, 
+                    Array.Empty<ParameterModifier>())
+                .Invoke(new object[] { _env, columns });
```
Comment (adamsitnik) [inline src/benchmarks/real-world/Microsoft.ML.Benchmarks/HashBench.cs]: I really don't like reflection, but I could not find any other way to get it working without introducing changes to ML.NET.

I was able to compile following code:

```cs
var info = _env.Transforms.Conversion.Hash("Bar", "Foo", numberOfBits: numberOfBits);
var dataView = _env.Data.LoadFromEnumerable(Array.Empty<RowImpl>(), _inRow.Schema);
var xf = info.Fit(dataView);
```

but I was getting errors like this:

```log
 System.ArgumentOutOfRangeException: Could not determine an IDataView type for member Schema
Parameter name: rawType
   at Microsoft.ML.Data.InternalSchemaDefinition.GetVectorAndItemType(String name, Type rawType, IEnumerable`1 attributes, Boolean& isVector, Type& itemType)
   at Microsoft.ML.Data.InternalSchemaDefinition.GetVectorAndItemType(MemberInfo memberInfo, Boolean& isVector, Type& itemType)
   at Microsoft.ML.Data.SchemaDefinition.Create(Type userType, Direction direction)
   at Microsoft.ML.Data.DataViewConstructionUtils.GetSchemaDefinition[TRow](IHostEnvironment env, DataViewSchema schema)
   at Microsoft.ML.Data.DataViewConstructionUtils.GetInternalSchemaDefinition[TRow](IHostEnvironment env, DataViewSchema schema)
   at Microsoft.ML.Data.DataViewConstructionUtils.CreateFromEnumerable[TRow](IHostEnvironment env, IEnumerable`1 data, DataViewSchema schema)
   at Microsoft.ML.Benchmarks.HashBench.InitMap[T](T val, DataViewType type, Int32 numberOfBits, ValueGetter`1 getter)
   at Microsoft.ML.Benchmarks.HashBench.SetupHashScalarString() in C:\Projects\performance\src\benchmarks\real-world\Microsoft.ML.Benchmarks\HashBench.cs:line 136
```

Comment (eerhardt) [reply src/benchmarks/real-world/Microsoft.ML.Benchmarks/HashBench.cs]: You can remove this by cherry-picking this change: [https://github.com/eerhardt/performance/commit/db30f57f91175610b3bfbfd30e788b245f03facd](https://github.com/eerhardt/performance/commit/db30f57f91175610b3bfbfd30e788b245f03facd)

---
In reply to: [341010904](https://github.com/dotnet/performance/pull/987#discussion_r341010904) [](ancestors = 341010904)

Comment (eerhardt) [reply src/benchmarks/real-world/Microsoft.ML.Benchmarks/HashBench.cs]: I've pushed the change to your branch.

**Comment by adamsitnik** [comment]
@eerhardt thank you for the review and the fix!

**Comment by BruceForstall** [comment]
@adamsitnik Now that the release version of ML.NET is out (1.4.0 instead of 1.4.0-preview2), does it make sense to update to that (and stick with it?)?

**Comment by adamsitnik** [comment]
> Now that the release version of ML.NET is out (1.4.0 instead of 1.4.0-preview2), does it make sense to update to that (and stick with it?)?

**Pull request merged**